cirq/circuits/circuit_operation.py

```diff
@@ -0,0 +1,384 @@
+        (or values) used when applying this operation as part of a larger
+        circuit.
+        """
+        return self._param_resolver
+
+    def __eq__(self, other) -> bool:
+        if not isinstance(other, type(self)):
+            return False
+        return (
+            self.circuit == other.circuit
+            and self.qubit_map == other.qubit_map
+            and self.measurement_key_map == other.measurement_key_map
+            and self.param_resolver == other.param_resolver
+            and self.repetitions == other.repetitions
+        )
+
+    # Methods for getting post-mapping properties of the contained circuit.
+
+    @property
+    def qubits(self) -> Tuple['cirq.Qid', ...]:
+        """Returns the qubits operated on by this object."""
+        ordered_qubits = ops.QubitOrder.DEFAULT.order_for(self.circuit.all_qubits())
+        return tuple(self._qubit_map.get(q, q) for q in ordered_qubits)
+
+    def _qid_shape_(self) -> Tuple[int, ...]:
+        return tuple(q.dimension for q in self.qubits)
+
+    def _measurement_keys_(self):
+        return {
+            self.measurement_key_map.get(key, key) for key in self.circuit.all_measurement_keys()
+        }
+
+    def _parameter_names_(self):
+        return {
+            name
+            for symbol in protocols.parameter_symbols(self.circuit)
+            for name in protocols.parameter_names(
+                protocols.resolve_parameters(symbol, self.param_resolver)
+            )
+        }
+
+    def _decompose_(self) -> 'cirq.OP_TREE':
+        result = self.circuit.unfreeze()
+        result = result.transform_qubits(lambda q: self.qubit_map[q])
+        if self.repetitions < 0:
+            result = result ** -1
+        result = protocols.with_measurement_key_mapping(result, self.measurement_key_map)
+        result = protocols.resolve_parameters(result, self.param_resolver)
+
+        return list(result.all_operations()) * abs(self.repetitions)
+
+    # Methods for string representation of the operation.
+
+    def __repr__(self):
+        base_repr = f'cirq.CircuitOperation({self.circuit!r})'
+        base_op = self.base_operation()
+
+        def dict_repr(d: Dict) -> str:
+            pairs = [f'    {proper_repr(k)}: {proper_repr(v)},' for k, v in sorted(d.items())]
+            return '\n'.join(['{'] + pairs + ['}'])
+
+        if self.qubit_map != base_op.qubit_map:
+            base_repr += f'.with_qubit_mapping({dict_repr(self.qubit_map)})'
+        if self.measurement_key_map != base_op.measurement_key_map:
+            base_repr += (
+                '.with_measurement_key_mapping(' + f'{dict_repr(self.measurement_key_map)})'
+            )
+        if self.param_resolver != base_op.param_resolver:
+            base_repr += f'.with_params({proper_repr(self.param_resolver)})'
+        if self.repetitions != 1:
+            base_repr += f'.repeat({self.repetitions})'
+        return base_repr
```

Comment:
This repr with a bunch of method calls seems a bit odd to me. What if we change the constructor to accept optional args for qubit_mapping, params, etc. and then can have one constructor call for the repr `cirq.CircuitOperation(..., qubit_mapping={...}, ...)`.

I think this might also simplify other code where were modify various mappings to just construct new instances instead of copy + mutate.